source/base/mpi.cc

```diff
@@ -812,7 +814,534 @@ namespace Utilities
+                                       &send_requests[index]);
+                AssertThrowMPI(ierr1);
+
+                // start to receive data
+                auto &recv_buffer = recv_buffers[index];
+                recv_buffer.resize(indices_i.size());
+                auto ierr2 = MPI_Irecv(recv_buffer.data(),
+                                       recv_buffer.size(),
+                                       MPI_UNSIGNED,
+                                       rank,
+                                       dict.offset + tag_delivery,
+                                       comm,
+                                       &recv_requests[index]);
+                AssertThrowMPI(ierr2);
+              }
+          }
+        }
+
+        void
+        clean_up_and_end_communication_with_dict(
+          std::vector<unsigned int> &array)
+        {
+          // clean up
+          {
+            auto ierr1 = MPI_Waitall(n_relevant_procs,
+                                     send_requests.data(),
+                                     MPI_STATUSES_IGNORE);
+            AssertThrowMPI(ierr1);
+
+            auto ierr2 = MPI_Waitall(n_relevant_procs,
+                                     recv_requests.data(),
+                                     MPI_STATUSES_IGNORE);
+            AssertThrowMPI(ierr2);
+
+            auto ierr3 = MPI_Wait(&barrier_request, MPI_STATUS_IGNORE);
+            AssertThrowMPI(ierr3);
+
+            for (auto &i : request_requests)
+              {
+                auto ierr = MPI_Wait(i.get(), MPI_STATUS_IGNORE);
+                AssertThrowMPI(ierr);
+              }
+          }
+
+          // unpack data
+          {
+            for (unsigned int i = 0; i < n_relevant_procs; i++)
+              for (unsigned int j = 0; j < recv_indices[i].size(); j++)
+                array[recv_indices[i][j]] = recv_buffers[i][j];
+          }
+        }
+      };
+
+      void
+      setup_dictionary(Dictionary &    dict,
+                       const IndexSet &owned_dofs,
+                       const MPI_Comm &comm)
+      {
+        unsigned int my_rank = this_mpi_process(comm);
+
+        // 1) setup dictionary and allocate memory
+        dict.reinit(owned_dofs, comm);
+
+
+#  ifdef DEBUG
+        for (auto &a : dict.array)
+          a = numbers::invalid_unsigned_int;
+#  endif
+
+        types::global_dof_index              dic_local_rececived = 0;
+        std::map<unsigned int, unsigned int> relevant_procs_map;
+
+        // 2) collect relevant processes and process local dict entries
+        {
+          std::set<unsigned int> relevant_procs;
+          for (auto i : owned_dofs)
+            {
+              unsigned int other_rank = dict.dof_to_dict_rank(i);
+              if (other_rank == my_rank)
+                {
+                  dict.array[i - dict.local_range.first] = my_rank;
+                  dic_local_rececived++;
+                }
+              else
+                relevant_procs.insert(other_rank);
```

Comment:
This is probably a micro optimization but given the fact that `owned_dofs` passes through indices in ascending order and `dof_to_dict_rank` is monotone as well, you could simply choose `std::vector<unsigned int>` as for the `relevant_procs` field and
```
if (relevant_procs.empty() || relevant_procs.back() != other_rank)
  relevant_procs.push_back(other_rank);
```
instead of this one. The `for (auto i : owned_dofs)` is probably also somewhat expensive because iterating over ranges, with additional checks whether to increment within a range or to increment the range, is a hell for the compiler to optimize. Nonetheless, I don't think it pays of to allocate a separate array `std::vector<types::global_dof_index>`, fill it via `IndexSet::fill_index_vector()`, and iterate over the vector, so keep that part, especially due to the simple code.